tokio/src/io/driver/registration.rs

```diff
@@ -232,7 +239,8 @@ cfg_io_readiness! {
             pin!(fut);
 
             crate::future::poll_fn(|cx| {
-                if self.handle.inner().is_none() {
+                let inner = self.handle.inner();
+                if inner.is_none() || inner.filter(|i| i.is_shutdown()).is_some() {
```

Comment:
I was surprised that you needed to add this logic to make the tests pass. I did some digging and it appears that it is needed because the multi-threaded runtimes parker holds a strong ref to the I/O driver, which means that wakers hold strong refs to the driver. This goes back to #3481.